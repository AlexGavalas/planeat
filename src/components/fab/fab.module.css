.container {
    position: fixed;
    bottom: 20px;
    left: 20px;
}

.button {
    max-width: fit-content;
}
