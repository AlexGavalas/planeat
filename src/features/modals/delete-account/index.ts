export { DeleteAccountModal } from './delete-account';
