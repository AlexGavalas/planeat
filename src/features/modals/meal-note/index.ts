export { MealNoteModal } from './meal-note';
