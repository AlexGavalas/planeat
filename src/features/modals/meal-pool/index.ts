export { MealPool } from './meal-pool';
