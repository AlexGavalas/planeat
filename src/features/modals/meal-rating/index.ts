export { MealRatingModal } from './meal-rating';
