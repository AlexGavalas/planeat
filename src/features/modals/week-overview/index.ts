export { WeekOverview } from './week-overview';
