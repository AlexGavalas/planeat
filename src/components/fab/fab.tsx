import {
    ActionIcon,
    Box,
    Button,
    type ButtonProps,
    Stack,
} from '@mantine/core';
import { useModals } from '@mantine/modals';
import { EditPencil, Plus, Running } from 'iconoir-react';
import { useTranslation } from 'next-i18next';
import { useCallback, useState } from 'react';
import { useQueryClient } from 'react-query';

import { ActivityModal } from '~features/modals/activity';
import { MeasurementModal } from '~features/modals/measurement';
import { useProfile } from '~hooks/use-profile';

import styles from './fab.module.css';

const buttonProps = {
    className: styles.button,
    size: 'sm',
    radius: 'xl',
    variant: 'filled',
} satisfies ButtonProps;

export const Fab = () => {
    const { t } = useTranslation();
    const modals = useModals();
    const { profile } = useProfile();
    const queryClient = useQueryClient();
    const [showMenu, setShowMenu] = useState(false);

    const toggleMenu = useCallback(() => {
        setShowMenu((prev) => !prev);
    }, []);

    const handleMeasurementSave = useCallback(async () => {
        await queryClient.resetQueries({ queryKey: ['bmi-timeline'] });
        await queryClient.resetQueries({ queryKey: ['measurements'] });
        await queryClient.resetQueries({
            queryKey: ['fat-percent-timeline'],
        });
    }, [queryClient]);

    const handleActivitySave = useCallback(async () => {
        await queryClient.resetQueries({ queryKey: ['activities-count'] });
        await queryClient.resetQueries({ queryKey: ['activities'] });
    }, [queryClient]);

    const handleAddMeasurement = useCallback(() => {
        if (!profile) {
            return;
        }

        toggleMenu();

        modals.openModal({
            title: t('add_measurement'),
            size: 'md',
            children: (
                <MeasurementModal
                    userId={profile.id}
                    onSave={handleMeasurementSave}
                />
            ),
        });
    }, [handleMeasurementSave, modals, profile, t, toggleMenu]);

    const handleAddActivity = useCallback(() => {
        if (!profile) {
            return;
        }

        toggleMenu();

        modals.openModal({
            title: t('add_activity'),
            size: 'md',
            children: (
                <ActivityModal
                    userId={profile.id}
                    onSave={handleActivitySave}
                />
            ),
        });
    }, [handleActivitySave, modals, profile, t, toggleMenu]);

    return (
        <Box className={styles.container}>
            {showMenu && (
                <Stack gap="sm" mb={10}>
                    <Button
                        {...buttonProps}
                        leftSection={<EditPencil />}
                        onClick={handleAddMeasurement}
                    >
                        {t('add_measurement')}
                    </Button>
                    <Button
                        {...buttonProps}
                        leftSection={<Running />}
                        onClick={handleAddActivity}
                    >
                        {t('add_activity')}
                    </Button>
                </Stack>
            )}
            <ActionIcon
                color="brand"
                size="xl"
                radius="xl"
                variant="filled"
                onClick={toggleMenu}
            >
                <Plus />
            </ActionIcon>
        </Box>
    );
};
