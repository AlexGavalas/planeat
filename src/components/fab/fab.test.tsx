import { ModalsProvider } from '@mantine/modals';
import { QueryClient, QueryClientProvider } from 'react-query';

import { renderWithUser } from '~test/utils';

import { Fab } from './fab';

jest.mock('next/router', () => jest.requireActual('next-router-mock'));

jest.mock('~hooks/use-profile', () => ({
    useProfile: jest.fn().mockReturnValue({
        profile: { id: '1' },
    }),
}));

describe('<Fab />', () => {
    it('renders initially closed', () => {
        const { container } = renderWithUser(<Fab />, {
            Wrapper: ({ children }) => (
                <QueryClientProvider client={new QueryClient()}>
                    <ModalsProvider>{children}</ModalsProvider>
                </QueryClientProvider>
            ),
        });

        expect(container).toMatchSnapshot();
    });
});
