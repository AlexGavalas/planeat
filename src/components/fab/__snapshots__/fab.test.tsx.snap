// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fab /> renders initially closed 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="container"
  >
    <button
      class="mantine-focus-auto mantine-active m-8d3f4000 mantine-ActionIcon-root m-87cf2631 mantine-UnstyledButton-root"
      data-size="xl"
      data-variant="filled"
      style="--ai-size: var(--ai-size-xl); --ai-radius: var(--mantine-radius-xl); --ai-bg: brand; --ai-hover: rgba(0, 0, 0, 1); --ai-color: var(--mantine-color-white); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
      type="button"
    >
      <svg
        color="currentColor"
        fill="none"
        height="1.5em"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="1.5em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 12h6m6 0h-6m0 0V6m0 6v6"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</div>
`;
